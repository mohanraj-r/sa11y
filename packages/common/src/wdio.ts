/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { BrowserObject, MultiRemoteBrowserObject } from 'webdriverio';
import { A11yConfig } from './axe';
import { ExceptionList } from './format';

export type WDIOBrowser = BrowserObject | MultiRemoteBrowserObject;

/**
 * Optional arguments passed to WDIO APIs
 * @param driver - WDIO {@link BrowserObject} instance navigated to the page to be checked. Created automatically by WDIO test runner. Might need to be passed in explicitly when other test runners are used.
 * @param scope - Element to check for accessibility found using [`browser.$(selector)`](https://webdriver.io/docs/selectors), defaults to the entire document.
 * @param rules - {@link A11yConfig} to be used for checking accessibility. Defaults to {@link recommended}
 */
export interface Options {
    driver: WDIOBrowser;
    scope?: Promise<WebdriverIO.Element>;
    rules?: A11yConfig;
    exceptionList?: ExceptionList;
}

export type AssertFunction = (opts: Partial<Options>) => void | Promise<void>;
