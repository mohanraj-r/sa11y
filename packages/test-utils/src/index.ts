/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    domWithA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    domWithVisualA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    htmlFileWithVisualA11yIssues,
    a11yIssuesCount,
    shadowDomID,
    videoURL,
} from './test-data';
export { beforeEachSetup, cartesianProduct, checkA11yError } from './utils';
export { setupWDIO, teardownWDIO } from './wdio';
