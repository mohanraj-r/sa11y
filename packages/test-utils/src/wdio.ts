/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { BrowserObject, remote, RemoteOptions } from 'webdriverio';

export const defaultJestTimeout = 5000; // Is there a way to get this dynamically ?

export async function setupWDIO(): Promise<void> {
    jest.setTimeout(defaultJestTimeout * 10); // Increase timeout for WebdriverIO tests
    if (!global.browser) global.browser = await getBrowser();
}

export async function teardownWDIO(): Promise<void> {
    jest.setTimeout(defaultJestTimeout);
    if (global.browser) {
        const browser = global.browser;
        // Note: deleteSession() doesn't clear the browser instance immediately.
        //  browser.status() hangs after deleteSession is called.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        global.browser = undefined;
        // TODO (fix): Jest warning: "A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --runInBand --detectOpenHandles to find leaks."
        await browser.deleteSession();
    }
}

export async function getBrowser(debug = false): Promise<BrowserObject> {
    // Ref: https://www.npmjs.com/package/@wdio/sync#using-webdriverio-as-standalone-package
    return await remote(getConfigWDIO(debug));
}

export function getConfigWDIO(debug = false): RemoteOptions {
    // TODO (refactor): Is there a way to reuse the config object from wdio.conf.js here ?
    //  https://webdriver.io/docs/typescript.html#typed-configuration
    // const config = require('../../../wdio.conf.js');
    return {
        runner: 'local',
        specs: ['__tests__/**/*.ts'],
        filesToWatch: ['src/**/*.ts'],
        maxInstances: debug ? 1 : 10,
        // TODO (chore): "capabilities" is an object in RemoteOptions but an array in Config. Clarify if this is expected.
        capabilities: {
            maxInstances: debug ? 1 : 5,
            browserName: 'chrome',
            'goog:chromeOptions': {
                // https://developers.google.com/web/updates/2017/04/headless-chrome)
                args: debug ? [] : ['--headless', '--disable-gpu'],
            },
        },
        // Additional list of node arguments to use when starting child processes
        execArgv: debug ? ['--inspect', '--trace-warnings'] : ['--unhandled-rejections=strict'],
        // Level of logging verbosity: trace | debug | info | warn | error | silent
        logLevel: debug ? 'debug' : 'error',
        outputDir: debug ? '/tmp/wdio/' : '',
        bail: debug ? 1 : 0,
        baseUrl: 'http://localhost',
        // Default timeout for all waitFor* commands.
        waitforTimeout: 10000,
        // Default timeout in milliseconds for request
        // if browser driver or grid doesn't send response
        connectionRetryTimeout: 120000,
        // Default request retries count
        connectionRetryCount: 3,
        services: ['chromedriver'],
        framework: 'mocha',
        reporters: ['spec'],
        // Options to be passed to Mocha.
        // See the full list at http://mochajs.org/
        // TODO (chore): file issue with WDIO for missing type def
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        //@ts-ignore mochaOpts doesn't exist in webdriverIO Config type
        mochaOpts: {
            // TypeScript setup
            // TODO (refactor): Can we reuse babel instead https://webdriver.io/docs/babel.html
            //  https://github.com/webdriverio/webdriverio/issues/5588
            require: ['ts-node/register'],
            ui: 'bdd',
            timeout: debug ? 600000000 : 60000,
        },
    };
}
