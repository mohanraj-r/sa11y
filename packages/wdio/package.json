{
    "name": "@sa11y/wdio",
    "version": "0.2.0-alpha.0",
    "description": "Accessibility testing API for WebdriverIO",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/wdio#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/wdio"
    },
    "keywords": [
        "accessibility",
        "automated test",
        "integration test",
        "Webdriver",
        "WebdriverIO",
        "axe"
    ],
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "scripts": {
        "test": "wdio run ../../wdio.conf.js",
        "test:debug": "DEBUG=true yarn test",
        "test:watch": "yarn test --watch"
    },
    "dependencies": {
        "@sa11y/common": "0.2.1-beta.0",
        "@sa11y/format": "0.2.2-beta.0",
        "@sa11y/preset-rules": "0.3.0-beta.0",
        "axe-core": "4.0.2"
    },
    "peerDependencies": {
        "webdriverio": ">=6.0.0"
    },
    "devDependencies": {
        "@sa11y/test-utils": "0.2.1-beta.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
