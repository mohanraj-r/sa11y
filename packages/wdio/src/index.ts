/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { assertAccessible, assertAccessibleSync, Options } from './wdio';
export { setup } from './setup';
