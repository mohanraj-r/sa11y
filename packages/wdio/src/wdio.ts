/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { BrowserObject } from 'webdriverio';
import { A11yConfig, recommended } from '@sa11y/preset-rules';
import { A11yError } from '@sa11y/format';
import { getViolations } from '@sa11y/common';

// TODO (refactor): Find a way to declare version into axe namespace
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
// eslint-disable-next-line import/namespace
export const axeVersion: string | undefined = axe.version;

/**
 * Return version of axe injected into browser
 */
export async function getAxeVersion(driver: BrowserObject): Promise<typeof axeVersion> {
    return driver.execute(() => {
        // TODO (refactor): Find a way to declare version into axe namespace
        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
        // @ts-ignore
        // eslint-disable-next-line import/namespace
        return typeof axe === 'object' ? axe.version : undefined;
    });
}

/**
 * Load axe source into browser if it is not already loaded and return version of axe
 */
export async function loadAxe(driver: BrowserObject): Promise<void> {
    if ((await getAxeVersion(driver)) !== axeVersion) {
        await driver.execute(axe.source);
    }

    if ((await getAxeVersion(driver)) !== axeVersion) {
        throw new Error('Unable to load axe');
    }
}

/**
 * Load and run axe in given WDIO instance and return the accessibility violations found.
 */
export async function runAxe(driver: BrowserObject, rules: A11yConfig = recommended): Promise<axe.Result[]> {
    await loadAxe(driver);

    // run axe inside browser and return violations
    return await driver.executeAsync((rules, done) => {
        axe.run(document, rules, (err: Error, results: axe.AxeResults) => {
            if (err) throw err;
            done(results.violations);
        });
    }, rules);
}

/**
 * Verify that the currently loaded page in the browser is accessible.
 * Throw an error with the accessibility issues found if it is not accessible.
 * Asynchronous version of {@link assertAccessibleSync}
 * @param driver - WDIO browser instance navigated to page to be checked
 * @param rules - a11y preset-rules to be used for checking accessibility
 */
export async function assertAccessible(
    driver: BrowserObject = browser,
    rules: A11yConfig = recommended
): Promise<void> {
    // TODO (feat): Add as custom commands to both browser for page level and elem
    //      https://webdriver.io/docs/customcommands.html
    const violations = await getViolations(() => runAxe(driver, rules));
    A11yError.checkAndThrow(violations);
}

/**
 * Verify that the currently loaded page in the browser is accessible.
 * Throw an error with the accessibility issues found if it is not accessible.
 * Synchronous version of {@link assertAccessible}
 * @param driver - WDIO browser instance navigated to page to be checked
 * @param rules - a11y preset-rules to be used for checking accessibility
 */
export function assertAccessibleSync(driver: BrowserObject = browser, rules: A11yConfig = recommended): void {
    // Note: https://github.com/webdriverio/webdriverio/tree/master/packages/wdio-sync#switching-between-sync-and-async
    driver.call(async () => {
        await assertAccessible(driver, rules);
    });
}
