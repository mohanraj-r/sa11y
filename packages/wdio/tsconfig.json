{
    "extends": "../../tsconfig.wdio.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist",
        "noEmitOnError": false
    },
    "include": ["src"],
    "references": [
        { "path": "../preset-rules" },
        { "path": "../format" },
        { "path": "../common" },
        { "path": "../test-utils" }
    ]
}
