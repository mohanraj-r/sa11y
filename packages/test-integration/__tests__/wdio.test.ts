/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { BrowserObject, remote } from 'webdriverio';
import { assertAccessibleSync } from '@sa11y/wdio';
import { htmlFileWithA11yIssues, htmlFileWithNoA11yIssues } from '@sa11y/test-utils';

const sync = require('@wdio/sync').default;
const defaultJestTimeout = 5000; // Is there a way to get this dynamically ?

let browser: BrowserObject;

beforeAll(async () => {
    jest.setTimeout(defaultJestTimeout * 100); // Increase timeout for WebdriverIO tests

    // Ref: https://www.npmjs.com/package/@wdio/sync#using-webdriverio-as-standalone-package
    // TODO (refactor): Is there a way to reuse the config object from wdio.conf.js here ?
    browser = await remote({
        runner: 'local',
        capabilities: {
            browserName: 'chrome',
            /* cSpell:disable */
            'goog:chromeOptions': {
                // https://developers.google.com/web/updates/2017/04/headless-chrome)
                args: ['--headless', '--disable-gpu'],
            },
            /* cSpell:enable */
        },
        logLevel: 'warn',
    });
});

afterAll(async () => {
    await browser.deleteSession();
    jest.setTimeout(defaultJestTimeout);
});

describe('integration test @sa11y/wdio in sync mode', () => {
    it('should throw error for html with a11y issues', () => {
        expect(browser).toBeTruthy();
        return sync(() => {
            browser.url(htmlFileWithA11yIssues);
            expect(() => assertAccessibleSync(browser)).toThrowErrorMatchingSnapshot();
        });
    });

    it('should not throw error for html with no a11y issues', () => {
        expect(browser).toBeTruthy();
        return sync(() => {
            browser.url(htmlFileWithNoA11yIssues);
            expect(() => assertAccessibleSync(browser)).not.toThrow();
        });
    });
});
