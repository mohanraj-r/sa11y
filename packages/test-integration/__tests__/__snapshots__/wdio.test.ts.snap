// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test @sa11y/wdio in sync mode should throw error for html with a11y issues 1`] = `
"6 accessibility issues found
 * (bypass) Page must have means to bypass repeated blocks: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/bypass

* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/document-title

* (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/html-has-lang

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/link-name

* (landmark-one-main) Document must have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/landmark-one-main

* (page-has-heading-one) Page must contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/page-has-heading-one"
`;

exports[`integration test @sa11y/wdio in sync mode should throw error for html with visual a11y issues 1`] = `
"2 accessibility issues found
 * (color-contrast) Elements must have sufficient color contrast: #upside-down,button
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/color-contrast

* (no-autoplay-audio) <video> or <audio> elements do not autoplay audio: audio
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/no-autoplay-audio

* (no-autoplay-audio) <video> or <audio> elements do not autoplay audio: video
	- Help URL: https://dequeuniversity.com/rules/axe/4.0/no-autoplay-audio"
`;
